src/Features/Core/Portable/PullMemberUp/AbstractCodeAndSolutionGenerator.cs

```diff
@@ -0,0 +1,336 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+
+namespace Microsoft.CodeAnalysis.PullMemberUp
+{
+    internal abstract class AbstractCodeAndSolutionGenerator : IPullMemberUpActionAndSolutionGenerator
+    {
+        private Dictionary<DocumentId, DocumentEditor> EditorMap { get; set; }
+
+        private Dictionary<ISymbol, IEnumerable<SyntaxNode>> SyntaxMap { get; set; }
+
+        private ICodeGenerationService CodeGenerationService { get; set; }
+
+        public async Task<Solution> GetSolutionAsync(
+            AnalysisResult result,
+            Document contextDocument,
+            CancellationToken cancellationToken = default)
```

Comment:
for internal code we don't use optional cancellation tokens.  otherwise people can easily forget to pass it along.